File: offer/20/JPhoebe.py
"""
定义栈的数据结构，请在该类型中实现一个能够得到栈中所含最小元素的min函数（时间复杂度应为O（1））。
"""

class Solution:
    stack = []
    min_list = []

    def push(self, node):
        # 每次压入数据栈数据与最小数据进行比较，yes: 存入数据栈、辅助栈
        self.stack.append(node)
        if len(self.min_list) < 1:
            self.min_list.append(node)
        else:
            if self.min_list[-1] > node:
                self.min_list.append(node)

    def pop(self):
        if self.stack is None:
            return None
        if self.stack[-1] == self.min_list[-1]:
            self.min_list.pop()
        value = self.stack.pop()
        return value

    def top(self):
        if self.stack is None:
            return None
        value = self.stack[-1]
        return value

    def min(self):
        if self.min_list is None:
            return None
        value = self.min_list[-1]
        return value

File: offer/59/jphoebe.java
import java.util.*;

/*
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
public class Solution {
    Stack<TreeNode> odd = new Stack<>();
    Stack<TreeNode> even = new Stack<>();
    int count = 0 , nextLineCount = 1, lineNumber = 1;
    ArrayList<ArrayList<Integer>> result = new ArrayList<>();
    ArrayList<Integer> line = new ArrayList<>();

    public ArrayList<ArrayList<Integer>> Print(TreeNode pRoot) {
        if (pRoot == null){
            return new ArrayList<>();
        }
        odd.add(pRoot);
        while (!odd.isEmpty() || !even.isEmpty() ){
            count ++;
            TreeNode node;
            if (lineNumber % 2 == 1){
                // 下一行偶数行 右遍历
                node = odd.pop();
                line.add(node.val);
                if (node.left != null){
                    even.add(node.left);
                }
                if (node.right != null){
                    even.add(node.right);
                }

                if (count == nextLineCount){
                    nextLineCount = even.size();
                    count = 0;
                    lineNumber ++;
                    result.add(line);
                    line = new ArrayList<>();
                }
            }else{
                // 下一行技术行 左遍历
                node = even.pop();
                line.add(node.val);
                if (node.right != null){
                    odd.add(node.right);
                }
                if (node.left != null){
                    odd.add(node.left);
                }

                if (count == nextLineCount){
                    nextLineCount = odd.size();
                    count = 0;
                    lineNumber ++;
                    result.add(line);
                    line = new ArrayList<>();
                }
            }

        }
        return result;
    }

}
